// vant-weapp/dist/mixins/transition.js
import { isObj } from '../common/utils';
const getClassNames = name => ({
  enter: `van-${name}-enter van-${name}-enter-active enter-class enter-active-class`,
  'enter-to': `van-${name}-enter-to van-${name}-enter-active enter-to-class enter-active-class`,
  leave: `van-${name}-leave van-${name}-leave-active leave-class leave-active-class`,
  'leave-to': `van-${name}-leave-to van-${name}-leave-active leave-to-class leave-active-class`
});
const nextTick = () => new Promise(resolve => setTimeout(resolve, 1000 / 30));

export default function (showDefaultValue) {
  return {
    data() {
      return {
        type: '',
        inited: false,
        display: false
      };
    },

    props: {
      customStyle: String,
      // @ts-ignore
      show: {
        type: Boolean,
        default: showDefaultValue
      },
      // @ts-ignore
      duration: {
        type: [Number, Object],
        default: 300
      },
      name: {
        type: String,
        default: 'fade'
      }
    },
    watch: {
      show: {
        handler: 'observeShow',
        immediate: true
      },
      duration: {
        handler: 'observeDuration',
        immediate: true
      }
    },

    beforeMount() {
      if (this.show) {
        this.enter();
      }
    },

    methods: {
      observeShow(value) {
        if (value) {
          this.enter();
        } else {
          this.leave();
        }
      },

      enter() {
        const {
          duration,
          name
        } = this;
        const classNames = getClassNames(name);
        const currentDuration = isObj(duration) ? duration.enter : duration;
        this.status = 'enter';
        Promise.resolve().then(nextTick).then(() => {
          this.checkStatus('enter');
          this.set({
            inited: true,
            display: true,
            classes: classNames.enter,
            currentDuration
          });
        }).then(nextTick).then(() => {
          this.checkStatus('enter');
          this.set({
            classes: classNames['enter-to']
          });
        }).catch(() => {});
      },

      leave() {
        const {
          duration,
          name
        } = this;
        const classNames = getClassNames(name);
        const currentDuration = isObj(duration) ? duration.leave : duration;
        this.status = 'leave';
        Promise.resolve().then(nextTick).then(() => {
          this.checkStatus('leave');
          this.set({
            classes: classNames.leave,
            currentDuration
          });
        }).then(() => setTimeout(() => this.onTransitionEnd(), currentDuration)).then(nextTick).then(() => {
          this.checkStatus('leave');
          this.set({
            classes: classNames['leave-to']
          });
        }).catch(() => {});
      },

      checkStatus(status) {
        if (status !== this.status) {
          throw new Error(`incongruent status: ${status}`);
        }
      },

      onTransitionEnd() {
        if (!this.show) {
          this.set({
            display: false
          });
          this.$emit('transitionEnd');
        }
      }

    }
  };
}
